[Java] Abstract class corner cases failing: arrays and enums

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/fory/issues) and found no similar issues.


### Version

OS: ArchLinux (6.16.7-zen1-1-zen)
Fory: 'org.apache.fory:fory-core:0.13.0-SNAPSHOT' (Sep 30)
Failing JDK: GraalVM 23.0.2+7.1 (build 23.0.2+7-jvmci-b01)

### Component(s)

Java

### Minimal reproduce step

Corner case 1: `AbstractObject[]`: https://github.com/gudzpoz/fory-native-image-reproducer/blob/abstract-array/app/src/main/java/org/example/App.java

```console
$ git clone https://github.com/gudzpoz/fory-native-image-reproducer.git
$ git checkout abstract-array
$ ./gradlew run
Caused by: org.apache.fory.exception.InsecureException: class org.example.App$AbstractObject is not registered, please check whether it's the type you want to serialize or a **vulnerability**. If safe, you should invoke `Fory#register` to register class,  which will have better performance by skipping classname serialization. If your env is 100% secure, you can also avoid this exception by disabling class registration check using `ForyBuilder#requireClassRegistration(false)`
```

Corner case 2: Abstract enums: (failing only under native images) https://github.com/gudzpoz/fory-native-image-reproducer/blob/abstract-enums/app/src/main/java/org/example/App.java

```console
$ git clone https://github.com/gudzpoz/fory-native-image-reproducer.git
$ git checkout abstract-enum
$ ./gradlew run # pass
$ ./gradlew nativeCompile
$ app/build/native/nativeCompile/test
java.lang.RuntimeException: Class class org.example.App$MyEnum is not registered
```

### What did you expect to see?

The tests should pass.

### What did you see instead?

Corner case 1: `AbstractObject[]`:

```console
Exception in thread "main" java.lang.ExceptionInInitializerError
Caused by: org.apache.fory.exception.InsecureException: class org.example.App$AbstractObject is not registered, please check whether it's the type you want to serialize or a **vulnerability**. If safe, you should invoke `Fory#register` to register class,  which will have better performance by skipping classname serialization. If your env is 100% secure, you can also avoid this exception by disabling class registration check using `ForyBuilder#requireClassRegistration(false)`
        at org.apache.fory.resolver.ClassResolver.createSerializer(ClassResolver.java:1220)
        at org.apache.fory.resolver.ClassResolver.getClassInfo(ClassResolver.java:1135)
        at org.apache.fory.resolver.ClassResolver.createSerializer0(ClassResolver.java:1269)
        at org.apache.fory.resolver.ClassResolver.lambda$ensureSerializersCompiled$14(ClassResolver.java:1803)
        at org.apache.fory.collection.ForyObjectMap.forEach(ForyObjectMap.java:302)
        at org.apache.fory.resolver.ClassResolver.ensureSerializersCompiled(ClassResolver.java:1796)
        at org.apache.fory.Fory.ensureSerializersCompiled(Fory.java:1606)
        at org.example.App.<clinit>(App.java:20)
```

Corner case 2: Abstract enums: (failing only under native images)

```console
Exception in thread "main" org.apache.fory.exception.SerializationException: java.lang.RuntimeException: Class class org.example.App$MyEnum is not registered
	at org.apache.fory.Fory.processSerializationError(Fory.java:360)
	at org.apache.fory.Fory.serialize(Fory.java:329)
	at org.apache.fory.Fory.serialize(Fory.java:277)
	at org.example.App.main(App.java:31)
	at java.base@25/java.lang.invoke.LambdaForm$DMH/sa346b79c.invokeStaticInit(LambdaForm$DMH)
Caused by: java.lang.RuntimeException: Class class org.example.App$MyEnum is not registered
	at org.apache.fory.resolver.TypeResolver.getSerializerClassFromGraalvmRegistry(TypeResolver.java:635)
	at org.apache.fory.resolver.ClassResolver.getSerializerClass(ClassResolver.java:862)
	at org.apache.fory.resolver.ClassResolver.getSerializerClass(ClassResolver.java:854)
	at org.apache.fory.resolver.ClassResolver.createSerializer(ClassResolver.java:1260)
	at org.apache.fory.resolver.ClassResolver.getOrUpdateClassInfo(ClassResolver.java:1182)
	at org.apache.fory.resolver.ClassResolver.getRawSerializer(ClassResolver.java:847)
	at org.example.AppForyRefCodec_0.<init>(AppForyRefCodec_0.java:38)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.methodhandles.Util_java_lang_invoke_MethodHandle.invokeInternal(Target_java_lang_invoke_MethodHandle.java:241)
	at java.base@25/java.lang.invoke.MethodHandle.invokeBasic(MethodHandle.java:102)
	at java.base@25/java.lang.invoke.MethodHandle.invokeBasic(MethodHandle.java:0)
	at java.base@25/java.lang.invoke.Invokers$Holder.invoke_MT(Invokers$Holder)
	at org.apache.fory.serializer.Serializers.newSerializer(Serializers.java:101)
	at org.apache.fory.resolver.ClassResolver.createSerializer(ClassResolver.java:1261)
	at org.apache.fory.resolver.ClassResolver.getOrUpdateClassInfo(ClassResolver.java:1182)
	at org.apache.fory.Fory.write(Fory.java:401)
	at org.apache.fory.Fory.serialize(Fory.java:323)
	... 3 more
```

### Anything Else?

This seems to fix the issues, but I'm not entirely sure how enums are serialized:

```diff
diff --git c/java/fory-core/src/main/java/org/apache/fory/resolver/ClassResolver.java w/java/fory-core/src/main/java/org/apache/fory/resolver/ClassResolver.java
index 8b9eca8d..b36cc1ea 100644
--- c/java/fory-core/src/main/java/org/apache/fory/resolver/ClassResolver.java
+++ w/java/fory-core/src/main/java/org/apache/fory/resolver/ClassResolver.java
@@ -1800,7 +1800,10 @@ public class ClassResolver extends TypeResolver {
                 createSerializer0(cls);
               }
               if (cls.isArray()) {
-                createSerializer0(TypeUtils.getArrayComponent(cls));
+                Class<?> component = TypeUtils.getArrayComponent(cls);
+                if (isSerializable(component)) {
+                  createSerializer0(component);
+                }
               }
             }
           });
diff --git c/java/fory-core/src/main/java/org/apache/fory/resolver/TypeResolver.java w/java/fory-core/src/main/java/org/apache/fory/resolver/TypeResolver.java
index 4db04ecf..ea62ab3b 100644
--- c/java/fory-core/src/main/java/org/apache/fory/resolver/TypeResolver.java
+++ w/java/fory-core/src/main/java/org/apache/fory/resolver/TypeResolver.java
@@ -444,6 +444,9 @@ public abstract class TypeResolver {
   public abstract ClassDef getTypeDef(Class<?> cls, boolean resolveParent);
 
   public final boolean isSerializable(Class<?> cls) {
+    if (cls.isEnum()) {
+      return true;
+    }
     if (ReflectionUtils.isAbstract(cls) || cls.isInterface()) {
       return false;
     }
```

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

--- a/integration_tests/graalvm_tests/src/main/java/org/apache/fory/graalvm/AbstractClassExample.java
+++ b/integration_tests/graalvm_tests/src/main/java/org/apache/fory/graalvm/AbstractClassExample.java
@@ -1,0 +1,251 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.fory.graalvm;
+
+import java.util.Arrays;
+import java.util.Objects;
+import org.apache.fory.Fory;
+import org.apache.fory.util.Preconditions;
+
+/**
+ * Test for abstract class corner cases in GraalVM native image. This tests the fix for issue #2695:
+ * Abstract enums (enums with abstract methods) and arrays of abstract types.
+ */
+public class AbstractClassExample {
+
+  // Abstract enum with abstract methods - each enum value is an anonymous inner class
+  public enum AbstractEnum {
+    VALUE1 {
+      @Override
+      public int getValue() {
+        return 1;
+      }
+    },
+    VALUE2 {
+      @Override
+      public int getValue() {
+        return 2;
+      }
+    };
+
+    public abstract int getValue();
+  }
+
+  // Abstract base class for testing abstract object arrays
+  public abstract static class AbstractBase {
+    public int id;
+
+    public abstract String getType();
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) return false;
+      AbstractBase that = (AbstractBase) o;
+      return id == that.id;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(id);
+    }
+  }
+
+  // Concrete implementation 1
+  public static class ConcreteA extends AbstractBase {
+    public String name;
+
+    @Override
+    public String getType() {
+      return "A";
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) return false;
+      if (!super.equals(o)) return false;
+      ConcreteA concreteA = (ConcreteA) o;
+      return Objects.equals(name, concreteA.name);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(super.hashCode(), name);
+    }
+  }
+
+  // Concrete implementation 2
+  public static class ConcreteB extends AbstractBase {
+    public long value;
+
+    @Override
+    public String getType() {
+      return "B";
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) return false;
+      if (!super.equals(o)) return false;
+      ConcreteB concreteB = (ConcreteB) o;
+      return value == concreteB.value;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(super.hashCode(), value);
+    }
+  }
+
+  // Container class that holds abstract enum and abstract array
+  public static class Container {
+    public AbstractEnum enumValue;
+    public AbstractEnum[] enumArray;
+    public AbstractBase[] baseArray;
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) return false;
+      Container container = (Container) o;
+      return enumValue == container.enumValue
+          && Arrays.equals(enumArray, container.enumArray)
+          && Arrays.equals(baseArray, container.baseArray);
+    }
+
+    @Override
+    public int hashCode() {
+      int result = Objects.hash(enumValue);
+      result = 31 * result + Arrays.hashCode(enumArray);
+      result = 31 * result + Arrays.hashCode(baseArray);
+      return result;
+    }
+  }
+
+  private static final Fory FORY;
+
+  static {
+    FORY =
+        Fory.builder()
+            .withName(AbstractClassExample.class.getName())
+            .registerGuavaTypes(false)
+            .build();
+    // Register enum type - abstract enums need to be registered
+    // The fix for issue #2695 ensures that registering an abstract enum
+    // also registers its inner classes (the anonymous enum value classes)
+    FORY.register(AbstractEnum.class);
+    // Register concrete types
+    FORY.register(ConcreteA.class);
+    FORY.register(ConcreteB.class);
+    FORY.register(Container.class);
+    // Register array types - the abstract component type should be handled correctly
+    FORY.register(AbstractBase[].class);
+    FORY.register(AbstractEnum[].class);
+    // Ensure serializers are compiled - this is where the fix for issue #2695 matters
+    FORY.ensureSerializersCompiled();
+  }
+
+  public static void main(String[] args) {
+    FORY.reset();
+
+    // Test abstract enum serialization
+    testAbstractEnum();
+
+    // Test abstract enum array serialization
+    testAbstractEnumArray();
+
+    // Test abstract object array serialization
+    testAbstractObjectArray();
+
+    // Test container with abstract types
+    testContainer();
+
+    System.out.println("AbstractClassExample succeed");
+  }
+
+  private static void testAbstractEnum() {
+    byte[] bytes1 = FORY.serialize(AbstractEnum.VALUE1);
+    AbstractEnum result1 = FORY.deserialize(bytes1, AbstractEnum.class);
+    Preconditions.checkArgument(result1 == AbstractEnum.VALUE1, "VALUE1 should match");
+    Preconditions.checkArgument(result1.getValue() == 1, "VALUE1.getValue() should be 1");
+
+    byte[] bytes2 = FORY.serialize(AbstractEnum.VALUE2);
+    AbstractEnum result2 = FORY.deserialize(bytes2, AbstractEnum.class);
+    Preconditions.checkArgument(result2 == AbstractEnum.VALUE2, "VALUE2 should match");
+    Preconditions.checkArgument(result2.getValue() == 2, "VALUE2.getValue() should be 2");
+  }
+
+  private static void testAbstractEnumArray() {
+    AbstractEnum[] array = new AbstractEnum[] {AbstractEnum.VALUE1, AbstractEnum.VALUE2};
+    byte[] bytes = FORY.serialize(array);
+    AbstractEnum[] result = FORY.deserialize(bytes, AbstractEnum[].class);
+    Preconditions.checkArgument(Arrays.equals(array, result), "Enum arrays should match");
+    Preconditions.checkArgument(result[0].getValue() == 1, "result[0].getValue() should be 1");
+    Preconditions.checkArgument(result[1].getValue() == 2, "result[1].getValue() should be 2");
+  }
+
+  private static void testAbstractObjectArray() {
+    ConcreteA a = new ConcreteA();
+    a.id = 1;
+    a.name = "test";
+
+    ConcreteB b = new ConcreteB();
+    b.id = 2;
+    b.value = 100L;
+
+    AbstractBase[] array = new AbstractBase[] {a, b};
+    byte[] bytes = FORY.serialize(array);
+    AbstractBase[] result = FORY.deserialize(bytes, AbstractBase[].class);
+
+    Preconditions.checkArgument(result.length == 2, "Array length should be 2");
+    Preconditions.checkArgument(result[0] instanceof ConcreteA, "result[0] should be ConcreteA");
+    Preconditions.checkArgument(result[1] instanceof ConcreteB, "result[1] should be ConcreteB");
+    Preconditions.checkArgument(result[0].equals(a), "result[0] should equal a");
+    Preconditions.checkArgument(result[1].equals(b), "result[1] should equal b");
+    Preconditions.checkArgument(
+        "A".equals(result[0].getType()), "result[0].getType() should be 'A'");
+    Preconditions.checkArgument(
+        "B".equals(result[1].getType()), "result[1].getType() should be 'B'");
+  }
+
+  private static void testContainer() {
+    ConcreteA a = new ConcreteA();
+    a.id = 10;
+    a.name = "containerTest";
+
+    ConcreteB b = new ConcreteB();
+    b.id = 20;
+    b.value = 200L;
+
+    Container container = new Container();
+    container.enumValue = AbstractEnum.VALUE1;
+    container.enumArray = new AbstractEnum[] {AbstractEnum.VALUE2, AbstractEnum.VALUE1};
+    container.baseArray = new AbstractBase[] {a, b};
+
+    byte[] bytes = FORY.serialize(container);
+    Container result = FORY.deserialize(bytes, Container.class);
+
+    Preconditions.checkArgument(container.equals(result), "Container should match");
+    Preconditions.checkArgument(
+        result.enumValue.getValue() == 1, "container.enumValue.getValue() should be 1");
+  }
+}

--- a/integration_tests/graalvm_tests/src/main/java/org/apache/fory/graalvm/Main.java
+++ b/integration_tests/graalvm_tests/src/main/java/org/apache/fory/graalvm/Main.java
@@ -40,6 +40,7 @@ public class Main {
     Benchmark.main(args);
     CollectionExample.main(args);
     ArrayExample.main(args);
+    AbstractClassExample.main(args);
     FeatureTestExample.main(args);
   }
 }

--- a/integration_tests/graalvm_tests/src/main/resources/META-INF/native-image/org.apache.fory/graalvm_tests/native-image.properties
+++ b/integration_tests/graalvm_tests/src/main/resources/META-INF/native-image/org.apache.fory/graalvm_tests/native-image.properties
@@ -32,6 +32,7 @@ Args=-H:+ReportExceptionStackTraces \
   org.apache.fory.graalvm.EnsureSerializerExample$CustomSerializer,\
   org.apache.fory.graalvm.CollectionExample,\
   org.apache.fory.graalvm.ArrayExample,\
+  org.apache.fory.graalvm.AbstractClassExample,\
   org.apache.fory.graalvm.Benchmark,\
   org.apache.fory.graalvm.FeatureTestExample,\
   org.apache.fory.graalvm.FeatureTestExample$PrivateConstructorClass,\

--- a/java/fory-core/src/main/java/org/apache/fory/resolver/ClassResolver.java
+++ b/java/fory-core/src/main/java/org/apache/fory/resolver/ClassResolver.java
@@ -1384,6 +1384,15 @@ public class ClassResolver extends TypeResolver {
       }
     }
 
+    // For enum value classes (anonymous inner classes of abstract enums),
+    // reuse the serializer from the declaring enum class
+    if (!cls.isEnum() && Enum.class.isAssignableFrom(cls) && cls != Enum.class) {
+      Class<?> enclosingClass = cls.getEnclosingClass();
+      if (enclosingClass != null && enclosingClass.isEnum()) {
+        return getSerializer(enclosingClass);
+      }
+    }
+
     if (extRegistry.serializerFactory != null) {
       Serializer serializer = extRegistry.serializerFactory.createSerializer(fory, cls);
       if (serializer != null) {
@@ -1469,6 +1478,14 @@ public class ClassResolver extends TypeResolver {
     }
     if (cls.isArray()) {
       return isSecure(TypeUtils.getArrayComponent(cls));
+    }
+    // For enum value classes (anonymous inner classes of abstract enums),
+    // check if the declaring enum class is secure
+    if (!cls.isEnum() && Enum.class.isAssignableFrom(cls) && cls != Enum.class) {
+      Class<?> enclosingClass = cls.getEnclosingClass();
+      if (enclosingClass != null && enclosingClass.isEnum()) {
+        return isSecure(enclosingClass);
+      }
     }
     if (fory.getConfig().requireClassRegistration()) {
       return Functions.isLambda(cls)
@@ -1958,19 +1975,35 @@ public class ClassResolver extends TypeResolver {
                 createSerializer0(cls);
               }
               if (cls.isArray()) {
-                // Also create serializer for the component type
-                createSerializer0(TypeUtils.getArrayComponent(cls));
+                // Also create serializer for the component type if it's serializable
+                Class<?> componentType = TypeUtils.getArrayComponent(cls);
+                if (isSerializable(componentType)) {
+                  createSerializer0(componentType);
+                }
               }
             }
             // Always ensure array class serializers and their component type serializers
             // are registered in GraalVM registry, since ObjectArraySerializer needs
             // the component type serializer at construction time
             if (cls.isArray() && GraalvmSupport.isGraalBuildtime()) {
-              // First ensure component type serializer is registered
+              // First ensure component type serializer is registered if it's serializable
               Class<?> componentType = TypeUtils.getArrayComponent(cls);
-              createSerializer0(componentType);
+              if (isSerializable(componentType)) {
+                createSerializer0(componentType);
+              }
               // Then register the array serializer
               createSerializer0(cls);
+            }
+            // For abstract enums, also create and store serializers for enum value classes
+            // so they are available at GraalVM runtime
+            if (cls.isEnum() && GraalvmSupport.isGraalBuildtime()) {
+              for (Object enumConstant : cls.getEnumConstants()) {
+                Class<?> enumValueClass = enumConstant.getClass();
+                if (enumValueClass != cls) {
+                  // Get serializer for the enum value class (will reuse the enum's serializer)
+                  getSerializer(enumValueClass);
+                }
+              }
             }
           });
       if (GraalvmSupport.isGraalBuildtime()) {

--- a/java/fory-core/src/main/java/org/apache/fory/resolver/TypeResolver.java
+++ b/java/fory-core/src/main/java/org/apache/fory/resolver/TypeResolver.java
@@ -481,6 +481,10 @@ public abstract class TypeResolver {
   public abstract ClassDef getTypeDef(Class<?> cls, boolean resolveParent);
 
   public final boolean isSerializable(Class<?> cls) {
+    // Enums are always serializable, even if abstract (enums with abstract methods)
+    if (cls.isEnum()) {
+      return true;
+    }
     if (ReflectionUtils.isAbstract(cls) || cls.isInterface()) {
       return false;
     }

--- a/java/fory-core/src/test/java/org/apache/fory/resolver/ClassResolverTest.java
+++ b/java/fory-core/src/test/java/org/apache/fory/resolver/ClassResolverTest.java
@@ -564,4 +564,106 @@ public class ClassResolverTest extends ForyTestBase {
         fory.getClassResolver().getSerializer(abs2Test.getClass()).getClass(),
         AbstractCustomSerializer.class);
   }
+
+  // Test enum with abstract methods (which makes the enum class abstract)
+  enum AbstractEnum {
+    VALUE1 {
+      @Override
+      public int getValue() {
+        return 1;
+      }
+    },
+    VALUE2 {
+      @Override
+      public int getValue() {
+        return 2;
+      }
+    };
+
+    public abstract int getValue();
+  }
+
+  @Test
+  public void testAbstractEnumIsSerializable() {
+    Fory fory = Fory.builder().withLanguage(Language.JAVA).requireClassRegistration(false).build();
+    ClassResolver classResolver = fory.getClassResolver();
+    // Abstract enums should be serializable
+    Assert.assertTrue(classResolver.isSerializable(AbstractEnum.class));
+    // The concrete enum value classes should also be serializable
+    Assert.assertTrue(classResolver.isSerializable(AbstractEnum.VALUE1.getClass()));
+    Assert.assertTrue(classResolver.isSerializable(AbstractEnum.VALUE2.getClass()));
+  }
+
+  @Test
+  public void testAbstractEnumSerialization() {
+    Fory fory = Fory.builder().withLanguage(Language.JAVA).requireClassRegistration(false).build();
+    // Serialize and deserialize abstract enum values
+    Assert.assertEquals(AbstractEnum.VALUE1, serDe(fory, AbstractEnum.VALUE1));
+    Assert.assertEquals(AbstractEnum.VALUE2, serDe(fory, AbstractEnum.VALUE2));
+    Assert.assertEquals(1, ((AbstractEnum) serDe(fory, AbstractEnum.VALUE1)).getValue());
+    Assert.assertEquals(2, ((AbstractEnum) serDe(fory, AbstractEnum.VALUE2)).getValue());
+  }
+
+  @Test
+  public void testAbstractObjectArraySerialization() {
+    Fory fory = Fory.builder().withLanguage(Language.JAVA).requireClassRegistration(false).build();
+    // Create an array of abstract type with concrete instances
+    AbsTest[] array = new AbsTest[2];
+    SubAbsTest item1 = new SubAbsTest();
+    item1.setF1(10);
+    item1.f2 = 100L;
+    Sub2AbsTest item2 = new Sub2AbsTest();
+    item2.setF1(20);
+    item2.f2 = 200L;
+    item2.f3 = "test";
+    array[0] = item1;
+    array[1] = item2;
+
+    AbsTest[] result = serDe(fory, array);
+    Assert.assertEquals(result.length, 2);
+    Assert.assertEquals(result[0].getF1(), 10);
+    Assert.assertEquals(((SubAbsTest) result[0]).f2, 100L);
+    Assert.assertEquals(result[1].getF1(), 20);
+    Assert.assertEquals(((Sub2AbsTest) result[1]).f2, 200L);
+    Assert.assertEquals(((Sub2AbsTest) result[1]).f3, "test");
+  }
+
+  @Test
+  public void testAbstractObjectArrayWithRegistration() {
+    Fory fory = Fory.builder().withLanguage(Language.JAVA).requireClassRegistration(true).build();
+    // Register the concrete types but not the abstract type
+    fory.register(SubAbsTest.class);
+    fory.register(Sub2AbsTest.class);
+    fory.register(AbsTest[].class);
+
+    AbsTest[] array = new AbsTest[2];
+    SubAbsTest item1 = new SubAbsTest();
+    item1.setF1(10);
+    item1.f2 = 100L;
+    Sub2AbsTest item2 = new Sub2AbsTest();
+    item2.setF1(20);
+    item2.f2 = 200L;
+    item2.f3 = "test";
+    array[0] = item1;
+    array[1] = item2;
+
+    AbsTest[] result = serDe(fory, array);
+    Assert.assertEquals(result.length, 2);
+    Assert.assertEquals(result[0].getF1(), 10);
+    Assert.assertEquals(result[1].getF1(), 20);
+  }
+
+  @Test
+  public void testAbstractEnumArraySerialization() {
+    Fory fory = Fory.builder().withLanguage(Language.JAVA).requireClassRegistration(false).build();
+    // Create an array of abstract enum type
+    AbstractEnum[] array = new AbstractEnum[] {AbstractEnum.VALUE1, AbstractEnum.VALUE2};
+
+    AbstractEnum[] result = serDe(fory, array);
+    Assert.assertEquals(result.length, 2);
+    Assert.assertEquals(result[0], AbstractEnum.VALUE1);
+    Assert.assertEquals(result[1], AbstractEnum.VALUE2);
+    Assert.assertEquals(result[0].getValue(), 1);
+    Assert.assertEquals(result[1].getValue(), 2);
+  }
 }

